src/sas/sascalc/calculator/geni.py

```diff
@@ -221,44 +206,60 @@ def _calc_Iqxy_magnetic(
     shape = qx.shape
     qx, qy = (np.asarray(v, 'd').flatten() for v in (qx, qy))
     Iq = np.zeros(shape=qx.shape, dtype='d')
+    M = np.array([mx, my, mz])
     #print("mag", [v.shape for v in (x, y, rho, vol, mx, my, mz)])
     _calc_Iqxy_magnetic_helper(
-        Iq, qx, qy, x, y, rho, vol, mx, my, mz,
-        cos_spin, sin_spin, dd, du, ud, uu)
+        Iq, qx, qy, x, y, rho, vol, M,
+        cos_spin, sin_spin, cos_phi, sin_phi, dd, du, ud, uu)
     return Iq.reshape(shape)
 
-@njit("(" + "f8[:], "*10 + "f8, "*6 + ")")
+@njit
+def orth(A, b): # A = 3 x n, and b_hat unit vector
+    return A - np.outer(b, b)@A
+
+@njit("(" + "f8[:], "*7 + "f8[:,::1], "+ "f8, "*8 + ")")
 def _calc_Iqxy_magnetic_helper(
-        Iq, qx, qy, x, y, rho, vol, mx, my, mz, cos_spin, sin_spin,
+        Iq, qx, qy, x, y, rho, vol, M, cos_spin, sin_spin, cos_phi, sin_phi,
         dd, du, ud, uu):
     # Process each qx, qy
     # Note: enumerating a pair is slower than direct indexing in numba
     for k in range(len(qx)):
         qxk, qyk = qx[k], qy[k]
-        # If q is near 0 then set px and py to zero.
-        # Note: norm is computed as a separate scalar so that the numba jit
-        # can figure out the proper type for perp even for q = 0
-        qsq = qxk**2 + qyk**2
-        norm = 1./qsq if qsq > 1e-16 else 0.
-        perp = norm*(qyk*mx - qxk*my)
-        px = perp*(qyk*cos_spin + qxk*sin_spin)
-        py = perp*(qyk*sin_spin - qxk*cos_spin)
-        pz = mz
 
-        ephase = vol*np.exp(1j*(qxk*x + qyk*y))
+        if abs(qxk) > 1.e-16 or abs(qyk) > 1.e-16:
+            norm = 1/np.sqrt(qxk**2 + qyk**2)
+            q_hat = np.array([qxk, qyk, 0]) * norm
+        else:
+            # For homogeneously magnetised disc Mperp can be associated to the
+            # magnetsation corrected for demag factorfield q->0, i.e. M-Nij M
+            # with Nij the demagnetisation tensor (Belleggia JMMM 263, L1, 2003).
+            q_hat = np.array([0.5, 0.5, 0])
```

Comment:
`q_hat` is not a unit vector. Is this intended?

Reply:
Right. The trace of np.outer in the function orth should be always 1.